### PR TITLE
[REVIEW] Add git commit hook for clang-format [skip ci]

### PR DESCRIPTION
**Pull request by codereport**
Add a git commit hook to automatically format C++/CUDA code when committing.

Resolves https://github.com/rapidsai/cudf/issues/4936

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by codereport** [comment]
image
Tested and it works.

**Comment by codereport** [comment]
> This is the configuration in the linked repo: https://github.com/doublify/pre-commit-clang-format/blob/master/.pre-commit-hooks.yaml
> 
> This doesn't include `.cu` / `.cuh` files, is that intended?

No it is not. Let me do a test to confirm that CUDA files are being missed. If so, I will look for a repo with CUDA files or might need to set up repo myself.

**Comment by kkraus14** [comment]
> No it is not. Let me do a test to confirm that CUDA files are being missed. If so, I will look for a repo with CUDA files or might need to set up repo myself.

Given the simplicity of the config and our need to possibly tweak it moving forward, I would suggest we set up a centralized repo of pre-commit hooks. @rapidsai/ops any opinions here?

**Comment by kkraus14** [comment]
@codereport to unblock short term I think you could just define the pre-commit hook inline and that should work.

**Comment by kkraus14** [comment]
Would like someone from the cpp side to review just to ensure we're capturing the expected extensions here :)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -19,5 +19,15 @@ repos:
...
                 args: ["--config=python/cudf/.flake8.cython"]
                 types: [cython]
+      - repo: local
+        hooks:
+              - id: clang-format
+                name: clang-format
+                description: Format files with ClangFormat.
+                entry: clang-format -i
+                language: system
+                files: \.(cu|cuh|h|hpp|cpp)$
```
Comment (davidwendt) [inline .pre-commit-config.yaml:29]: The strings source has a couple of `.inl` files that are included by `.cuh` headers.
Looks like they did not get reformatted with the great clang reformat so I'm guessing the python script may need to be updated too.

Comment (codereport) [reply .pre-commit-config.yaml:29]: Ok, I will setup a follow up PR to adjust git commit hook and cuDF sources at the same time.

**Comment by harrism** [comment]
How does one enable this pre commit hook?

**Comment by kkraus14** [comment]
> How does one enable this pre commit hook?

You should be able to follow the instructions here: https://github.com/rapidsai/cudf/blob/branch-0.14/CONTRIBUTING.md#code-formatting

**Comment by harrism** [comment]
Looks like those need to be updated to mention clang-format.

**Comment by codereport** [comment]
> Looks like those need to be updated to mention clang-format.

With my `conda` environment, `pre-commit` was already installed (I assume this is the case for many people). So all i needed to do was run:
```
pre-commit install
```
and it was set up. 

I will update the documentation.

**Pull request merged**